x-pack/plugins/ingest_manager/server/services/epm/fields/field.ts

```diff
@@ -43,45 +49,133 @@ export interface Field {
-      fields[key] = newField;
-      if (newField.fields) {
-        processFields(newField.fields);
+      // check child fields further down the tree
+      if (groupField.fields) {
+        expandFields(groupField.fields);
+      }
+      // Replace the original field in the array with the new one
+      fields[key] = groupField;
+    } else {
+      // even if this field doesn't have dots to expand, its child fields further down the tree might
+      if (field.fields) {
+        expandFields(field.fields);
       }
     }
   });
 }
+/**
+ * dedupFields takes the given fields and merges sibling fields with the
+ * same name together.
+ * These can result from expandFields when the input contains dotted field
+ * names that share parts of their hierarchy.
+ */
+function dedupFields(fields: Fields): Fields {
+  const dedupedFields: Fields = [];
+  fields.forEach(field => {
+    const found = dedupedFields.find(f => {
+      return f.name === field.name;
+    });
+    if (found) {
+      if (found.type === 'group' && field.type === 'group' && found.fields && field.fields) {
+        found.fields = dedupFields(found.fields.concat(field.fields));
+      } else {
+        // only 'group' fields can be merged in this way
+        throw new Error(
+          "Can't merge fields " + JSON.stringify(found) + ' and ' + JSON.stringify(field)
+        );
+      }
+    } else {
+      if (field.fields) {
+        field.fields = dedupFields(field.fields);
+      }
+      dedupedFields.push(field);
+    }
+  });
+  return dedupedFields;
+}
+
+/** validateAliasFields takes the given fields and verifies that all fields of type
+ * alias point to existing fields.
+ *
+ * Invalid alias fields are silently removed.
+ */
+
+function validateAliasFields(fields: Fields, allFields: Fields): Fields {
+  const validatedFields: Fields = [];
+
+  fields.forEach(field => {
+    if (field.type === 'alias') {
+      if (field.path && getField(allFields, field.path.split('.'))) {
+        validatedFields.push(field);
+      }
+    } else {
+      validatedFields.push(field);
+    }
+    if (field.fields) {
+      field.fields = validateAliasFields(field.fields, allFields);
+    }
+  });
+  return validatedFields;
+}
+
+const getField = (fields: Fields, pathNames: string[]): Field | undefined => {
+  if (!pathNames.length) return undefined;
+  // get the first rest of path names
+  const [name, ...restPathNames] = pathNames;
+  for (const field of fields) {
+    if (field.name === name) {
+      // check field's fields, passing in the remaining path names
+      if (field.fields && field.fields.length > 0) {
+        return getField(field.fields, restPathNames);
+      }
+      // no nested fields to search, but still more names - not found
+      if (restPathNames.length) {
+        return undefined;
+      }
+      return field;
+    }
+  }
+  return undefined;
+};
+
+export function processFields(fields: Fields): Fields {
+  expandFields(fields);
```

Comment:
can we make this more functional by not mutating the `fields `array like this?

Reply:
This is a left-over of the old `processFields` implementation. Would it be acceptable to change it in a follow-up PR?

Reply:
Yes, please.

Reply:
https://github.com/elastic/kibana/issues/60487